### ISSUE TITLE
3. BERT 언어모델 소개

### ISSUE THREAD

**Issue by pseeej** [sangmandu/SangSangPlus#84]
## CLS token 안에 정보가 축약되어 있다?
- CLS token은 문장의 시작이라고 생각했는데...
- 이 token을 이용하여 classification을 많이 함.
- **여러 개 문장의 정보를 축약**하고 있음. 단지 start의 의미가 아님
- **BERT는 양방향**이기 때문에 뒤의 문장의 정보까지 다 가질 수 있음
- pre-train될 때 CLS의 중요성보다, **fine-tuning에서의 output**에서 더 큰 의미를 가지는 것 같음

## token_str로 나오는 token들의 출처는 어디인가?
![image](https://user-images.githubusercontent.com/49185035/135042352-ba8ac5be-2753-4c21-a229-873b5d190040.png) ![image](https://user-images.githubusercontent.com/49185035/135042278-51ff9260-c856-424e-957a-2d94e63bba97.png)

**Comment by sangmandu** [comment]
## 형태소 tokenizing의 결과가 매번 다르게 나오는 것 같은데, 명확한 방법이 없는 것인가?

- 형태소는 문장에서 의미를 가지고 있는 가장 작은 단위.
- 형태소 단위는 답이 여러 개 나오는 것이 아니라, 딱 지정이 되어있는 것.
- 단지 **model의 성능 차이**에 따라 결과가 다르게 나오는 것으로 추정.

- _있습니다_ 의 경우... _있_ 과 _습니다_...로 추정

## BertWordPieceTokenizer의 train은 BERT를 학습시키는 것은 아니고, corpus를 tokenization하는 Byte Pair Encoding을 train하는 건가요?
- 맞습니다... **corpus를 tokenization하는 Byte Pair Encoding을 train**입니다...

## BERT의 pre-train 단계?
- mask된 단어의 embedding vector를 찾는 것. embedding vector를 만드는 과정
- 두 문장이 이어지는지 안 이어지는지 확인
- 두 개 이상을 예측해야 할 때는 그냥 token끼리 비교해서 정답을 찾아내는 것
- **pre-train은 self-supervised learning**임. data를 많이 쓰기 위해서... labeling을 모두 해주기 어려워서.
- 그래서 우리는 pre-train된 것만 가지고 fine-tuning을 진행하는 것인가.

## Mask LM의 정답과 비교하는 것은 "언제"...?
- token으로 만드는 순간에는 정답을 다 가지고 있는데 의도적으로 MASK로 가리는 것.
- 따라서 정답을 미리 갖고 있기 때문에 마지막에 loss를 계산할 때 비교하는 것이라고 생각.
- labeling을 따로 진행한 것은 아니지만 정답이 주어져있기 때문에 self-supervised learning이 아닌가
- embedding vector를 구하고 나서 원래 단어와의 차이를 계산하는 것?
- embedding layer층에서 뽑아낼 때?
- pre-training을 동시에 진행하는데 NSP의 정답은 [CLS] token에서 진행하고 layer 하나를 추가로 거친다
- MASK LM을 하기 위해서는 token의 index로 다시 한 번 나와야 하는데... 그 과정에서?

< 희락님 >
- 마지막 embedding 거치고 나온 vector에 weight 곱함으로써 layer층 더함.
- 거기서 softmax와 argmax 수행해서 원래 mask 씌우기 전의 문장과 비교.
- 다음 문장을 예측하는 것은 또 새로운 Fully-connected Layer를 사용
- Mask LM에서는 하나의 layer를 더 추가
![image](https://user-images.githubusercontent.com/49185035/135050467-0cb0c508-a73f-4468-b7c2-6fbadad34ad8.png)

**Comment by sangmandu** [comment]
## 질문 token_str로 나오는 token들의 출처는 어디인가? 🟢

## 답변
![image](https://user-images.githubusercontent.com/49185035/135042352-ba8ac5be-2753-4c21-a229-873b5d190040.png) ![image](https://user-images.githubusercontent.com/49185035/135042278-51ff9260-c856-424e-957a-2d94e63bba97.png)
- bert-base-multilingual-cased model을 사용함으로써 여기에 포함되어있는 token들을 사용했을 것
- wiki 문서를 이용한 pre-train을 통해 나온 단어들일 것

**Comment by sangmandu** [comment]
## [질문] BertWordPieceTokenizer의 train은 BERT를 학습시키는 것은 아니고, corpus를 tokenization하는 Byte Pair Encoding을 train하는 건가요?

## [답변]
- 맞습니다... **corpus를 tokenization하는 Byte Pair Encoding을 train**입니다...

## [질문] BERT의 pre-train 단계?


## [답변]
- mask된 단어의 embedding vector를 찾는 것. embedding vector를 만드는 과정
- 두 문장이 이어지는지 안 이어지는지 확인
- 두 개 이상을 예측해야 할 때는 그냥 token끼리 비교해서 정답을 찾아내는 것
- **pre-train은 self-supervised learning**임. data를 많이 쓰기 위해서... labeling을 모두 해주기 어려워서.
- 그래서 우리는 pre-train된 것만 가지고 fine-tuning을 진행하는 것인가.

**Comment by sangmandu** [comment]
## [질문] Mask LM의 정답과 비교하는 것은 "언제"...?

## [답변]
- token으로 만드는 순간에는 정답을 다 가지고 있는데 의도적으로 MASK로 가리는 것.
- 따라서 정답을 미리 갖고 있기 때문에 마지막에 loss를 계산할 때 비교하는 것이라고 생각.
- labeling을 따로 진행한 것은 아니지만 정답이 주어져있기 때문에 self-supervised learning이 아닌가
- embedding vector를 구하고 나서 원래 단어와의 차이를 계산하는 것?
- embedding layer층에서 뽑아낼 때?
- pre-training을 동시에 진행하는데 NSP의 정답은 [CLS] token에서 진행하고 layer 하나를 추가로 거친다
- MASK LM을 하기 위해서는 token의 index로 다시 한 번 나와야 하는데... 그 과정에서?

## [답변] 희락
- 마지막 embedding 거치고 나온 vector에 weight 곱함으로써 layer층 더함.
- 거기서 softmax와 argmax 수행해서 원래 mask 씌우기 전의 문장과 비교.
- 다음 문장을 예측하는 것은 또 새로운 Fully-connected Layer를 사용
- Mask LM에서는 하나의 layer를 더 추가
![image](https://user-images.githubusercontent.com/49185035/135050467-0cb0c508-a73f-4468-b7c2-6fbadad34ad8.png)

**Comment by sangmandu** [comment]
## [질문] BERT의 pre-train 단계? 🟢

## [답변]
- mask된 단어의 embedding vector를 찾는 것. embedding vector를 만드는 과정
- 두 문장이 이어지는지 안 이어지는지 확인
- 두 개 이상을 예측해야 할 때는 그냥 token끼리 비교해서 정답을 찾아내는 것
- **pre-train은 self-supervised learning**임. data를 많이 쓰기 위해서... labeling을 모두 해주기 어려워서.
- 그래서 우리는 pre-train된 것만 가지고 fine-tuning을 진행하는 것인가.
